### PR TITLE
[Qwen3.8-Flash-Next] Remove torch.compile for NVIDIA implementation

### LINKED ISSUE #54688: [RFC] Remove torch.compile dependency from Qwen3.8-Flash-Next NVIDIA path
## Motivation

The NVIDIA implementation of Qwen3.8-Flash-Next (`Qwen4Exp` internally), introduced in #53896, still applies `support_torch_compile` to the main model and MTP classes. CUDA graph capture does not require this model-level AOT compilation, and the model-specific operations are increasingly implemented as explicit vLLM custom ops.

## Kernel list

`✅` means that the PR body contains measured performance results. Unit tests, accuracy evaluations, benchmark claims without data, and empty benchmark sections are not counted.

| Kernel | PR | Microbenchmark | E2E benchmark |
|---|---|:---:|:---:|
| QSA uniform decode/spec-decode indexer | #54513 | ✅ | ✅ |
| QSA varlen prefill indexer | #54513 | ✅ | ✅ |
| QSA block-index expansion | #54513 | ✅ | ✅ |
| PLE n-gram ID generation | #54517 | ✅ | ✅ |
| PLE varlen dilated convolution + SiLU/residual accumulation | #54517 | ✅ | ✅ |
| PLE convolution-state writeback | #54517 | ✅ | ✅ |
| PLE three-RMSNorm gate + convolution-input preparation | #54517 | ✅ | ✅ |
| MTP embedding residual add + first grouped Gemma RMSNorm | #54687 | ✅ | ✅ |
| PLE outer residual-add fusion | — | — | — |
| QSA sparse-attention + sigmoid output-gate epilogue | — | — | — |

### Eager / compiled baseline

```mermaid
flowchart TB
    E0["Token / multimodal embeddings<br/>[T, H]"] --> E1["Expand to four HC streams<br/>[T, 4H]"]

    subgraph EB["Backbone: 48 × Qwen4ExpDecoderLayer"]
        E1 --> E2["Decoder layer i<br/>i = 1...48"]
        E2 --> E3{"i = 2?"}
        E3 -- Yes --> E4["PLE sublayer<br/>torch n-gram ops → embedding lookup → KV projection<br/>three RMSNorms + gate → dense packing + F.conv1d + SiLU/state<br/>hidden_states + ple_output"]
        E3 -- No --> E5
        E4 --> E5["Attention HyperConnection<br/>grouped RMSNorm → down/injection GEMM → SiLU<br/>up GEMM → sigmoid gate reduction"]
        E5 --> E6{"Attention type"}
        E6 -- "36 linear-attention layers" --> E7["Gated DeltaNet<br/>linear-attention path"]
        E6 -- "Layers 4, 8, ..., 48" --> E8["QSA full attention<br/>QKV+gate projection → Q/K norm+RoPE<br/>shared indexer → sparse attention<br/>attn_output × sigmoid(output_gate) → O projection"]
        E7 --> E9
        E8 --> E9["MLP HyperConnection<br/>injection combine → grouped RMSNorm → down/injection GEMM<br/>SiLU → up GEMM → sigmoid gate reduction"]
        E9 --> E10["MLP / 512-expert MoE block"]
        E10 --> E11{"i < 48?"}
        E11 -- Yes --> E2
    end

    E11 -- No --> E12["Backbone final HyperConnection mixer<br/>consume final pending combine → grouped RMSNorm<br/>low-rank gate → collapse four streams"]
    E12 --> E13["Target sample hidden [T, H]<br/>→ LM head / target logits"]
    E12 --> E14["Backbone multi-stream hidden [T, 4H]"]
    E14 --> E15["MTP input preparation<br/>token embedding norm + projection<br/>hidden norm + projection<br/>embedding broadcast add → flatten → grouped RMSNorm"]
    E15 --> E16["MTP draft decoder layer × 1<br/>full-attention HyperConnection + QSA + MoE"]
    E16 --> E17["MTP final HyperConnection mixer<br/>collapse four streams"]
    E17 --> E18["Draft sample hidden [T, H]<br/>→ LM head / draft logits"]

    classDef eager fill:#ffe0b2,stroke:#e65100,color:#000
    class E0,E1,E2,E4,E5,E7,E8,E9,E10,E12,E13,E14,E15,E16,E17,E18 eager
```

### Ideal explicit / fused path

```mermaid
flowchart TB
    F0["Token / multimodal embeddings<br/>[T, H]"] --> F1["Four-stream HC state<br/>[T, 4H]"]

    subgraph FB["Backbone: 48 × Qwen4ExpDecoderLayer"]
        F1 --> F2["Decoder layer i<br/>i = 1...48"]
        F2 --> F3{"i = 2?"}
        F3 -- Yes --> F4["Fused PLE sublayer<br/>fused n-gram ID generation<br/>embedding lookup + KV projection<br/>fused three RMSNorms + gate + outer residual initialization<br/>varlen dilated depthwise conv + SiLU + state write + residual accumulation"]
        F3 -- No --> F5
        F4 --> F5["Fused Attention HyperConnection<br/>pending combine + grouped RMSNorm<br/>down/injection GEMM + SiLU epilogue<br/>up GEMM + sigmoid gate-reduction epilogue"]
        F5 --> F6{"Attention type"}
        F6 -- "36 linear-attention layers" --> F7["Gated DeltaNet<br/>explicit linear-attention kernels"]
        F6 -- "Layers 4, 8, ..., 48" --> F8["Fused QSA full attention<br/>packed QKV+gate projection<br/>fused Q/K norm + RoPE + cache write<br/>shape-specialized prefill/decode indexer<br/>sparse attention + sigmoid output-gate epilogue<br/>O projection"]
        F7 --> F9
        F8 --> F9["Fused MLP HyperConnection<br/>injection combine + grouped RMSNorm<br/>down/injection GEMM + SiLU epilogue<br/>up GEMM + sigmoid gate-reduction epilogue"]
        F9 --> F10["Fused MLP / MoE kernels"]
        F10 --> F11{"i < 48?"}
        F11 -- Yes --> F2
    end

    F11 -- No --> F12["Fused backbone final HC mixer<br/>pending combine + grouped RMSNorm<br/>low-rank projections + gate reduction<br/>produce sample and multi-stream hidden"]
    F12 --> F13["Target sample hidden [T, H]<br/>→ LM head / target logits"]
    F12 --> F14["Backbone multi-stream hidden [T, 4H]"]
    F14 --> F15["Fused MTP input<br/>embedding/hidden projections<br/>broadcast residual add + first grouped RMSNorm"]
    F15 --> F16["MTP draft decoder layer × 1<br/>reuse fused full-attention HC, QSA, and MoE paths"]
    F16 --> F17["Fused MTP final HC mixer<br/>produce next multi-stream state and sample hidden"]
    F17 --> F18["Draft sample hidden [T, H]<br/>→ LM head / draft logits"]

    classDef fused fill:#e8f5e9,stroke:#2e7d32,color:#000
    class F0,F1,F2,F4,F5,F7,F8,F9,F10,F12,F13,F14,F15,F16,F17,F18 fused
```

## CC List

cc @ZJY0516 @gau-nernst

### PR DESCRIPTION
**Pull request by gau-nernst**
## Purpose

Fixes #54688
- Remove `@support_torch_compile` as model decorator
- Remove all custom op registrations, which were meant to bypass torch.compile inspection: `qwen4_exp_compute_ple_ngram_ids`, `qwen4_exp_ple_short_conv`, `qwen4_exp_qsa_with_output`
- Eager break insertion points (**IMPORTANT**): 

Another benefit of removing torch.compile is that now we can load FP8 ckpt on a single GB300. Previously, torch.compile attempts to autotune the embedding op, causing an extra ngram table tensor that consumes ~50GB -> OOM. Removing torch.compile naturally removes this activation peak.

## Test Plan

Model | MTP tokens | GSM8K | MMMU-Pro
-- | -- | -- | --
BF16 | 0 | 96.85% | 77.23%
BF16 | 3 | 96.66% | 77.57%
FP8 | 0 | 96.74% | 76.99%
FP8 | 3 | 96.82% | 75.61%

## E2E perf result

Model | Concurrency | Output tok/s | Throughput delta | TTFT ms | TPOT ms |
|---|---:|---:|---:|---:|---:|
| BF16 | 1 | 303.6 → 288.9 | -4.83% | 160.62 → 162.63 | 3.309 → 3.288 |
| BF16 | 4 | 759.8 → 742.6 | -2.26% | 195.19 → 188.14 | 5.020 → 5.259 |
| BF16 | 16 | 1582.2 → 1663.9 | +5.16% | 206.47 → 202.19 | 9.469 → 9.252 |
| BF16 | 64 | 3052.7 → 3123.3 | +2.31% | 231.19 → 222.05 | 20.354 → 20.173 |
| FP8 | 1 | 240.5 → 232.9 | -3.14% | 219.70 → 217.63 | 4.093 → 4.149 |
| FP8 | 4 | 642.8 → 631.8 | -1.72% | 238.66 → 238.27 | 5.858 → 6.150 |
| FP8 | 16 | 1345.7 → 1390.0 | +3.30% | 252.93 → 252.33 | 11.280 → 10.990 |
| FP8 | 64 | 2076.9 → 2066.3 | -0.51% | 308.32 → 309.84 | 30.071 → 30.199 |

Generally the results look noisy, so I don't think removing torch.compile right now is significantly worse. Future manual fusion will further improve perf.

## Test Result

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/model_executor/layers/mamba/gdn/qwen_gdn_linear_attn.py
Comment (ZJY0516) [inline vllm/model_executor/layers/mamba/gdn/qwen_gdn_linear_attn.py:2000]: These can be removed

Comment (gau-nernst) [reply vllm/model_executor/layers/mamba/gdn/qwen_gdn_linear_attn.py:2000]: Still needed for existing Qwen models i think

**Review by ZJY0516 — APPROVED** [review]

**Comment by de1tydev** [comment]
A field data point for the ~50 GB figure, from unified-memory hardware where the failure mode is worse than an OOM.

**Setup**: 2× DGX Spark (GB10, sm_121), TP2 across two nodes, vLLM nightly `e962733e`, `nvidia/Qwen3.8-Flash-Next-NVFP4`, MTP `k=5`, `gpu_memory_utilization=0.835`, `max_model_len=524288`. Standing config is compile off — `{"mode": 0, "cudagraph_mode": "FULL_DECODE_ONLY"}` — because compilation had already bitten us once on day 0.

On 2026-09-06 we A/B'd it back on as one rung of a performance ladder: `{"mode": 3, "cudagraph_mode": "FULL_AND_PIECEWISE"}`, everything else identical to a configuration that had just completed a full benchmark matrix. Weight loading was normal (~10 min). Once compilation started, **both hosts went unreachable at the same moment** and stayed that way for over two hours:

- ICMP kept responding;
- TCP connect to 22 and 8888 still completed;
- but the sshd handshake never finished and the API never answered — including retries with a 180 s timeout.

Recovery required power-cycling both machines.

On GB10 the CPU and GPU share one memory pool, so vLLM's 0.835 utilization already reserves roughly 100 GB per node. An extra ~50 GB ngram table for the embedding autotune does not land as a clean CUDA OOM there — it pushes the *host* into thrash, and userspace stops making progress while the kernel stays up. That is why there is no traceback to attach: no error, no crash, no core. Our supervisor did not fire either, because both its liveness checks needed the host to be able to run bash.

Two things this implies for anyone on unified-memory parts, which I think supports merging this:

1. The failure is silent and unattended-hostile. A compile-related OOM on a discrete-GPU box gets you an exception and a restart; here it gets you two dead nodes and no signal.
2. Any autotune or compile step whose peak allocation is a meaningful fraction of total system memory needs to account for host memory on these SKUs, not just free VRAM.

Removing `@support_torch_compile` for this path takes the whole class of failure off the table for us. Thanks for pushing it through.

**Comment by gau-nernst** [comment]
@de1tydev Glad it helps you!
